src/lib/project-fetcher-hoc.jsx

```diff
@@ -0,0 +1,137 @@
+            // debugger;
+            if (
+                props.projectId !== '' &&
+                props.projectId !== null &&
+                typeof props.projectId !== 'undefined'
+            ) {
+                this.props.setInitialProjectId(props.projectId);
+            }
+        }
+        componentDidUpdate (prevProps) {
+            if (prevProps.projectHost !== this.props.projectHost) {
+                storage.setProjectHost(this.props.projectHost);
+            }
+            if (prevProps.assetHost !== this.props.assetHost) {
+                storage.setAssetHost(this.props.assetHost);
+            }
+            if (this.props.isFetchingProjectWithId && !prevProps.isFetchingProjectWithId) {
+                this.fetchProject(this.props.reduxProjectId, this.props.projectState);
+            }
+        }
+        fetchProject (projectId, projectState) {
+            return storage
+                .load(storage.AssetType.Project, projectId, storage.DataFormat.JSON)
+                .then(projectAsset => {
+                    if (projectAsset) {
+                        this.props.doneFetchingProjectData(projectAsset.data, projectState);
+                    }
+                })
+                .then(() => {
+                    if (projectId !== defaultProjectId) {
+                        // if not default project, register a project load event
+                        analytics.event({
+                            category: 'project',
+                            action: 'Load Project',
+                            label: projectId,
+                            nonInteraction: true
+                        });
+                    }
+                })
+                .catch(err => {
+                    log.error(err);
+                });
+        }
+        render () {
+            const {
+                /* eslint-disable no-unused-vars */
+                assetHost,
+                doneFetchingProjectData: doneFetchingProjectDataProp,
+                intl,
+                projectHost,
+                projectId,
+                projectState,
+                reduxProjectId,
+                setInitialProjectId: setInitialProjectIdProp,
+                /* eslint-enable no-unused-vars */
+                isFetchingProjectWithId: isFetchingProjectWithIdProp,
+                ...componentProps
+            } = this.props;
+            return (
+                <WrappedComponent
+                    fetchingProject={isFetchingProjectWithIdProp}
+                    {...componentProps}
+                />
+            );
+        }
+    }
+    ProjectFetcherComponent.propTypes = {
+        assetHost: PropTypes.string,
+        doneFetchingProjectData: PropTypes.func,
+        intl: intlShape.isRequired,
+        isFetchingProjectWithId: PropTypes.bool,
+        projectHost: PropTypes.string,
+        projectId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
+        projectState: PropTypes.string,
```

Comment:
This would be a good place to document what this state is. `PropTypes.oneOf([list of states here])`. You could enumerate them in the reducer file and import them to keep things current, and that would also tell the reader where to look for how they're used.

Reply:
That's a cool idea, doing it!